cxx/pclean/pclean_lib.cc

```diff
@@ -0,0 +1,43 @@
+// Copyright 2024
+// Apache License, Version 2.0, refer to LICENSE.txt
+
+#include "irm.hh"
+#include "pclean/csv.hh"
+#include "pclean/pclean_lib.hh"
+
+T_observations translate_observations(
+    const DataFrame& df, const T_schema &schema) {
+  T_observations obs;
+
+  for (const auto& col : df.data) {
+    const std::string& col_name = col.first;
+    if (!schema.contains(col_name)) {
+      printf("Schema does not contain %s, skipping ...\n", col_name.c_str());
+      continue;
+    }
+
+    const T_relation& trel = schema.at(col_name);
+    size_t num_domains = std::visit([&](const auto &r) {
+      return r.domains.size();}, trel);
+
+    for (size_t i = 0; i < col.second.size(); ++i) {
+      const std::string& val = col.second[i];
+      if (val.empty()) {
+        // Don't incorporate missing values.
+        // TODO(thomaswc): Allow the user to specify other values that mean
+        // missing data.  ("missing", "NA", "nan", etc.).
+        continue;
+      }
+
+      std::vector<std::string> entities;
+      for (size_t j = 0; j < num_domains; ++j) {
+        // Give every row it's own universe of unique id's.
```

Comment:
Could you add a TODO to initialize this with a sample from the generative model instead (where only the Record class will have a unique ID per row)?

Reply:
You are using the word "unobserved" in a different way than I would use it.

For me, the schema creates the relations, and then the data comes in, and after all the data is in, we can mark a relation as observed or unobserved depending on whether it had any observations.

If I had to guess, you are maybe using "unobserved" to mean something "a relation for which we will need to initialize and maintain a latent state"?

But just to be clear, every observation here is for a noisy relation generated in the second part of PCleanSchemaHelper::make_hirm_schema -- those are the only ones that correspond to CSV column names.  None of those noisy relations will need to have a hidden latent state.  Even if somehow all of their data turned out to be missing from the CSV file, there are no other relations that depend on them.

Reply:
> For me, the schema creates the relations, and then the data comes in, and after all the data is in, we can mark a relation as observed or unobserved depending on whether it had any observations.

Ok, I think I see. The observed relations should be exactly those defined by the `observe` statement, so we can declare a relation to be "observed" if it appears in that statement, right? I wouldn't expect there to be data for relations not defined by `observe`, and if a relation defined by `observe` doesn't have any observations in the data, I think that's a user error. Do you agree?

> 
> If I had to guess, you are maybe using "unobserved" to mean something "a relation for which we will need to initialize and maintain a latent state"?

Assuming the observed relations are those defined by the `observe` statement (and those which have observed data) I think these are equivalent -- everything that isn't observed (which should be the base relations of noisy relations), we have to infer.

> 
> But just to be clear, every observation here is for a noisy relation generated in the second part of PCleanSchemaHelper::make_hirm_schema -- those are the only ones that correspond to CSV column names. None of those noisy relations will need to have a hidden latent state. Even if somehow all of their data turned out to be missing from the CSV file, there are no other relations that depend on them.

Reply:
> > For me, the schema creates the relations, and then the data comes in, and after all the data is in, we can mark a relation as observed or unobserved depending on whether it had any observations.
> 
> Ok, I think I see. The observed relations should be exactly those defined by the `observe` statement, so we can declare a relation to be "observed" if it appears in that statement, right? 

Well, again, I think the best/safest/most intuitive thing to do is to call a relation observed iff we see data for it.  So in line with that, I would mark relations as is_observed inside the incorporate_observations function.

But since we don't do that currently (is_observed is marked in load_observations), I've done the next best thing and declared the relations created from the "observe" clause in the schema as is_observed in this pull request.

> I wouldn't expect there to be data for relations not defined by `observe`, and if a relation defined by `observe` doesn't have any observations in the data, I think that's a user error. Do you agree?

Not quite.  If I have a schema and a csv file, and then I drop a column from the csv file, I wouldn't call it user error to run pclean on the combination, even though there would be a `observe` relation without any observations.  I think it would be quite easy to support that use case, so we probably should (while perhaps agreeing that supporting it isn't the most important priority for a research codebase).

> 
> > If I had to guess, you are maybe using "unobserved" to mean something "a relation for which we will need to initialize and maintain a latent state"?
> 
> Assuming the observed relations are those defined by the `observe` statement (and those which have observed data) I think these are equivalent -- everything that isn't observed (which should be the base relations of noisy relations), we have to infer.
> 
> > But just to be clear, every observation here is for a noisy relation generated in the second part of PCleanSchemaHelper::make_hirm_schema -- those are the only ones that correspond to CSV column names. None of those noisy relations will need to have a hidden latent state. Even if somehow all of their data turned out to be missing from the CSV file, there are no other relations that depend on them.

Reply:
I would also add that this conversation now has almost nothing to do with this pull request, and should probably be moved to a different channel (like a meeting or github issue).

Reply:
> 
> But just to be clear, every observation here is for a noisy relation generated in the second part of PCleanSchemaHelper::make_hirm_schema -- those are the only ones that correspond to CSV column names. None of those noisy relations will need to have a hidden latent state. Even if somehow all of their data turned out to be missing from the CSV file, there are no other relations that depend on them.

Could you add a TODO to sample the non-index domains from a CRP prior instead of assuming each entry is a unique entity? (We might want to rethink this more substantially for Model 7, i.e. whether it makes sense to initialize the entities as part of the data ingestion process or elsewhere. Other inferred values, e.g. latent values of a noisy relation, are initialized during `incorporate`, which might make more sense).

Reply:
Added a TODO to discuss and consider other options.

Reply:
> I would also add that this conversation now has almost nothing to do with this pull request, and should probably be moved to a different channel (like a meeting or github issue).

For me, this conversation was very relevant, and essential in allowing me to verify that this PR does the right thing, so I appreciate you taking the time to clarify.

Reply:
> But since we don't do that currently (is_observed is marked in load_observations), I've done the next best thing and declared the relations created from the "observe" clause in the schema as is_observed in this pull request.
> 
I was going to file a Github issue, but I think we should leave this as-is. Downstream in HIRM `!is_observed` implies the value is latent and needs to be inferred, so setting `is_observed` according to the "observe" clause gives the right semantics.

> > I wouldn't expect there to be data for relations not defined by `observe`, and if a relation defined by `observe` doesn't have any observations in the data, I think that's a user error. Do you agree?
> 
> Not quite. If I have a schema and a csv file, and then I drop a column from the csv file, I wouldn't call it user error to run pclean on the combination, even though there would be a `observe` relation without any observations. I think it would be quite easy to support that use case, so we probably should (while perhaps agreeing that supporting it isn't the most important priority for a research codebase).
> 
That's fine, I was thinking it made sense to support a more limited notion of valid inputs before potentially supporting something more general, but I agree it doesn't matter much here.